examples/t4rec_paper_experiments/t4r_paper_repro/transf_exp_main.py

```diff
@@ -182,6 +185,69 @@ def main():
 
         log_aot_metric_results(training_args.output_dir, results_avg_time)
 
+    # Mimic the inference by manually computing recall@10 using the evaluation data
+    # of the last time-index.
+    eval_path = os.path.join(
+        data_args.data_path,
+        str(
+            data_args.final_time_window_index,
+        ).zfill(data_args.time_window_folder_pad_digits),
+        "test.parquet" if training_args.eval_on_test_set else "valid.parquet",
+    )
+    prediction_data = pd.read_parquet(eval_path)
+    # Extract label
+    labels = prediction_data["sess_pid_seq"].apply(lambda x: x[-1]).values
+
+    # Truncate input sequences up to last item - 1 to mimic the inference
+    def mask_last_interaction(x):
+        return list(x[:-1])
+
+    list_columns = schema.select_by_tag("list").column_names
+    for col in list_columns:
+        prediction_data[col] = prediction_data[col].apply(mask_last_interaction)
+    # Get top-10 predictions
+    test_loader = NVTabularDataLoader.from_schema(
+        schema,
+        Dataset(prediction_data),
+        training_args.per_device_eval_batch_size,
+        max_sequence_length=training_args.max_sequence_length,
+        shuffle=False,
+    )
+    trainer.test_dataloader = test_loader
+    topk_preds = trainer.predict(test_loader).predictions[0]
+    # Compute recall@10
+    recall_10 = recall(topk_preds, labels)
+    logger.info(f"Recall@10 of manually masked test data = {str(recall_10)}")
+    output_file = os.path.join(training_args.output_dir, "eval_results_over_time.txt")
+    with open(output_file, "a") as writer:
+        writer.write(f"\n***** Recall@10 of simulated inference  = {recall_10} *****\n")
+    # Verify that the recall@10 from train.evaluate() matches the recall@10 calculated manually
+    if not isinstance(input_module.masking, t4r.masking.PermutationLanguageModeling):
+        # TODO fix inference discrepancy for permutation language modeling
+        assert np.isclose(recall_10, results_over_time[2]["eval_/next-item/recall_at_10"], rtol=0.1)
+
+
+def recall(predicted_items: np.ndarray, real_items: np.ndarray) -> float:
+    idx = 0
+    recalls = np.zeros(len(predicted_items), dtype=np.float64)
+    for real, pred in zip(real_items, predicted_items):
+
+        real = real[real > 0]
+        pred = pred[pred > 0]
```

Comment:
I understand `pred` is a 2D tensor with the top-k predicted ids, and eventually we might have a predicted item 0. Not sure has this op `pred[pred > 0]` will work in such case. I think it is better to align the filter with the `real`, (e.g. `pred[real > 0]`